contracts/optimize/OptimizedStorageManager.sol

```diff
@@ -0,0 +1,126 @@
+import "../StorageHelper.sol";
+import "../StorageSlotSelfDestructable.sol";
+
+contract OptimizedStorageManager {
+    uint internal constant SLOTLIMIT = 220; 
+    mapping(bytes32 => bytes32) public keyToContract;
+    
+    function _put(
+        bytes32 key,
+        bytes memory data,
+        uint256 value
+    ) internal {
+
+        bytes32 metadata = keyToContract[key];
+
+        if (!SlotHelper.isInSlot(metadata)){
+            address addr = SlotHelper.bytes32ToAddr(metadata);
+            // Notify :No need to delete metadata here, because it will be rewritten later
+            if (addr != address(0x0)) {
+                // remove the KV first if it exists
+                StorageSlotSelfDestructable(addr).destruct();
+            }
+        }
+    
+        if (data.length > SLOTLIMIT){
+            // store in new contract
+            keyToContract[key] = SlotHelper.addrToBytes32(StorageHelper.putRaw(data, value));
+        }else{
+            // store in slot
+            keyToContract[key] = SlotHelper.putRaw(key,data);
+            
+        }
+        
+    }
+
+    function _put2(
+        bytes32 key,
+        bytes memory data,
+        uint256 value
+    ) internal {
+        bytes32 metadata = keyToContract[key];
+        address addr = SlotHelper.bytes32ToAddr(metadata);
+        if (addr != address(0x0)) {
+            // remove the KV first if it exists
+            StorageSlotSelfDestructable(addr).destruct();
+        }
+
+        if (data.length > SLOTLIMIT){
+            // store in new contract
+            keyToContract[key] = SlotHelper.addrToBytes32(StorageHelper.putRaw2(key , data, value));
+        }else{
+            // store in slot
+            keyToContract[key] = SlotHelper.putRaw(key,data);
+        }
+
+    }
+
+    function _get(bytes32 key) internal view returns (bytes memory, bool) {
+        bytes32 metadata = keyToContract[key];
+        address addr = SlotHelper.bytes32ToAddr(metadata);
+
+        if (SlotHelper.isInSlot(metadata)){
+            bytes memory res  = SlotHelper.getRaw(key, metadata);
+            return (res,true);
+        }else{
+            return StorageHelper.getRaw(addr);
+        }
+
+    }
+
+    function _filesize(bytes32 key) internal view returns(uint){
+        bytes32 metadata = keyToContract[key];
+        address addr = SlotHelper.bytes32ToAddr(metadata);
+
+        if (metadata == bytes32(0)){
+            return 0;
+        }else if (SlotHelper.isInSlot(metadata)){
+            return SlotHelper.decodeLen(metadata);
+        }else{
+            (uint size, )= StorageHelper.sizeRaw(addr);
+            return size;
+        }
+    }
+
+    function _whereStore(bytes32 key) internal view returns(uint){
+        bytes32 metadata = keyToContract[key];
+        address addr = SlotHelper.bytes32ToAddr(metadata);
+
+        if (metadata == bytes32(0)){
+            return 0;
+        }else if (SlotHelper.isInSlot(metadata)){
+            return 1;
```

Comment:
I would suggest using constants instead of magic numbers.

Reply:
https://github.com/QuarkChain/evm-large-storage/pull/1/commits/4cb226647a5f56e77c05b2b09a50d23fc3300316